### DISCUSSION TITLE
Mixing Parallel and Sequential modes

### DISCUSSION THREAD

**Discussion by gofford** [ropensci/targets#543] [Help] [answered]
When working on a 16 CPU machine, I'm curious as to whether it's possible to mix and match multisession (future) and sequential execution modes within a single pipeline. I have scoured the documentation and have yet to figure it out, so wanted to seek feedback.

Consider this scenario:

* I have a multi-step graph
* Some steps are independent and can be executed in parallel safely. Let's say that they can use 1 CPU each in a multisession context, so we can execute 16 in parallel.
* Some steps are _very_ long running and execute in parallel under the hood. I'd like each of _these_ nodes to be executed in sequence, each using 16 cores, once the initial parallel jobs are completed.
* When the long-running jobs are completed, subsequent post-processing jobs can then be run in parallel again.

I hope this makes sense. Is this something that can be achieved in the current iteration of `targets`?

**Comment by wlandau** [comment, marked as answer]
So it sounds like we are talking about:

1. Parallelism *among* targets: multiple targets run simultaneously, versus
2. Parallelism *within* targets: an individual target runs multiple parallel processes.

(1) is explained at https://books.ropensci.org/targets/hpc.html. For (2), you can write custom code within a target to launch multiple parallel processes:

```r
tar_target(long_task, furrr::future_map(...))
```

If you already use `tar_make_future()`, this approach launches parallel workers within parallel workers, which may overburden the available cores in some cases. So in some cases, you could set `deployment = "main"` in `tar_target()`. This ensures the target launches from the main process instead of from an among-targets worker. It's not fitting for all types of pipelines through because it does block the main process until `long_task` finishes.

For this kind of nuanced parallelism, I highly recommend `future::plan(future.callr::callr)` instead of `future::plan(future::multisession)`. The latter sets up a PSOCK cluster in advance, which is basically a set of persistent workers. With the `callr` backend, there is no need to create all the workers up front, so the nesting is more efficient. And `callr` is generally more reliable than `parallel` because `callr` does not rely on sockets, connections, or forking.

**Reply by gofford** [reply]
Makes sense. I think this'll take a bit of experimenting on my part. It's one thing to have all jobs just share a shared pool of resources, but I find it a bit confusing how to mix/match.

Consider this toy schematic:

```
A1 -> B1 -> C1

A2 -> B2 -> C2

A3 -> B3 -> C3

AN -> BN -> CN
```

Where `N = 16` jobs. Essentially, what I'm trying to figure out is how to distribute these tasks such that `A` can execute in parallel (1 thread each), the `B` tasks each run one at a time (each using 16 threads), and the following `C` tasks go back to using 1 thread each. This is a very simplistic representation. In reality, the A tasks are data-load and pre-processing, the B tasks are `LightGBM` model fits (with `nthreads = 16` set, so fit in parallel out of the box) and the `C` tasks are post-processing and export jobs.

Would setting the `deployment = "main"` on the B jobs in this case ensure that only a single job is run, which can use 16 threads in the main session, with the workers idle in the background until needed? Is there any mechanism which stops the other jobs launching at the same time, or would they be queued on the `main` session?

Thanks for the help!

**Reply by wlandau** [reply]
In a way, `nthreads = 16` with `deployment = "main"` is almost equivalent to `nthreads = 1` with `deployment = "worker"` in terms of efficiency. In both, you are distributing the same physical volume of work equally over the same 16 workers. That's assuming `N = 16`, there is exactly 1 thread per core, all the A and C targets run super fast, and none of the B targets are already up to date. There are advantages of `nthreads = 16` with `deployment = "main"`: it will have higher occupancy if some of the B targets are up to date and can be skipped, and it is more resilient if the pipeline is interrupted. (It is more likely that some of the B targets will already have completed and can be skipped when the pipeline resumes). The main disadvantage of `nthreads = 16` with `deployment = "main"` is that it blocks other targets from launching while a B target is running, but that won't cost you in terms of practical efficiency because a single LightGBM model will use all 16 cores. If you want all the A targets to run, *then* all the B targets, *then* all the C targets, then you can set `priority` equal to 1, 0.5, and 0 respectively in `tar_target()`. Otherwise, `targets` will automatically traverse the DAG as described at https://books.ropensci.org/targets/hpc.html#transient-workers.

**Reply by wlandau** [reply]
> Would setting the deployment = "main" on the B jobs in this case ensure that only a single job is run, which can use 16 threads in the main session, with the workers idle in the background until needed?

With [transient workers](https://books.ropensci.org/targets/hpc.html#transient-workers), each target gets its own worker, and the worker terminates when the target finishes. So if you use `tar_make_future(workers = 16)`, then there will be no other workers running when a B target runs. If you were using [persistent workers](https://books.ropensci.org/targets/hpc.html#persistent-workers) with `tar_make_clustermq(workers = 16)`, on the other hand, then I would recommend deployment = "worker" with nthreads = 1 because otherwise all those 16 persistent workers would be idling on 16 cores while the main process also tries to use those same 16 cores for a single model.

**Reply by gofford** [reply]
Excellent, this makes sense @wlandau. Thank you. I'll explore this in more detail tomorrow and update this thread with  progress and feedback.

**Reply by tarensanders** [reply]
> For (2), you can write custom code within a target to launch multiple parallel processes:
> 
> ```r
> tar_target(long_task, furrr::future_map(...))
> ```

Can I clarify how you might do this if you are not mapping over values for the parallelism *within* targets? 

Take tuning a random forest, for example. In caret, you can set the models to tune in parallel like so (from [here](https://topepo.github.io/caret/parallel-processing.html)):

```r
library(doParallel)
cl <- makePSOCKcluster(5)
registerDoParallel(cl)

## All subsequent models are then run in parallel
model <- train(y ~ ., data = training, method = "rf")

## When you are done:
stopCluster(cl)
```

But it seems like you are warning against this, because it is creating workers within workers.

Say I had multiple random forests to train as the `B` jobs in @gofford's example, and I have 16 threads available. I don't actually care about running the B jobs one by one, since (as I think you're trying to establish above) running 16 threads on `B1` *then* 16 on `B2` is not actually faster than 8 threads each on `B1` and `B2` at the same time. Can I specify to targets to use as many threads as possible on `B` jobs if I'm not providing a `_map`-type function?

**Reply by wlandau** [reply]
> Can I clarify how you might do this if you are not mapping over values for the parallelism within targets?

It's the same principle, but with with everything from `cl <- makePSOCKcluster(5)` through `stopCluster(cl)` as custom code. So instead of 

```r
tar_target(long_task, furrr::future_map(...))
```

you would have

```r
run_in_cluster <- function(training) {
  registerDoParallel(cl)
  on.exit(stopCluster(cl))
  model <- train(y ~ ., data = training, method = "rf")
}

tar_target(long_task, run_in_cluster(training))
```

> Can I specify to targets to use as many threads as possible on B jobs if I'm not providing a _map-type function?

The threads/processes within a target are completely controlled by custom code you write in the target's command. There is no built-in way to control that through other arguments such as `workers` or `priority`.

**Reply by tarensanders** [reply]
Thank you, @wlandau. That makes complete sense. Very much appreciated.
